source/Engine.cpp

```diff
@@ -1930,113 +1930,133 @@ void Engine::HandleMouseClicks()
 // one adds any visuals that are created directly to the main visuals list. If
 // this is multi-threaded in the future, that will need to change.
 void Engine::DoCollisions(Projectile &projectile)
-{
-	// The asteroids can collide with projectiles, the same as any other
-	// object. If the asteroid turns out to be closer than the ship, it
-	// shields the ship (unless the projectile has a blast radius).
-	Point hitVelocity;
-	double closestHit = 1.;
-	shared_ptr<Ship> hit;
+{	
+	bool hasHit = true;
+	// Keep track of which ships this projectile has directly impacted this frame.
+	set<const Body *> hits;
 	const Government *gov = projectile.GetGovernment();
-	
-	// If this "projectile" is a ship explosion, it always explodes.
-	if(!gov)
-		closestHit = 0.;
-	else if(projectile.GetWeapon().IsPhasing() && projectile.Target())
-	{
-		// "Phasing" projectiles that have a target will never hit any other ship.
-		shared_ptr<Ship> target = projectile.TargetPtr();
-		if(target)
+	while(projectile.Penetrations() >= 0 && hasHit)
```

Comment:
You can:

- Rename `hasHit` to `allowMultipleHits` or something better.
- Turn this loop into a `do {} while` to avoid the akward `= false;` at the beginning of the loop.

Reply:
>Rename `hasHit` to `allowMultipleHits` or something better.

`allowMultipleHits` makes less sense to me, as what we want to do is break out of the loop if there's nothing left to it, hence the `hasHit` name.